File: mobile/test/data/sync/merge_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:english_helper_app/data/sync/merge.dart';

const t1 = '2026-08-01T00:00:00.000Z';
const t2 = '2026-08-02T00:00:00.000Z';

SyncRecord local(String id, String updated, String? synced) =>
    SyncRecord(id: id, updatedAt: updated, syncedAt: synced, data: {'id': id});

SyncRecord remote(String id, String updated) =>
    SyncRecord(id: id, updatedAt: updated, syncedAt: null, data: {'id': id});

void main() {
  test('규칙 1: 미동기 항목은 푸시하고 풀이 건드리지 않는다', () {
    final r = planMerge(local: [local('a', t1, null)], remote: []);
    expect(r.toPush.map((e) => e.id), ['a']);
    expect(r.toDeleteLocal, isEmpty);
    expect(r.toWriteLocal, isEmpty);
  });

  test('규칙 1: 서버에 더 오래된 사본이 있어도 미동기가 이긴다', () {
    final r = planMerge(local: [local('a', t2, null)], remote: [remote('a', t1)]);
    expect(r.toPush.map((e) => e.id), ['a']);
    expect(r.toWriteLocal, isEmpty);
  });

  test('규칙 2: 로컬이 더 최신이면 푸시', () {
    final r = planMerge(local: [local('a', t2, t1)], remote: [remote('a', t1)]);
    expect(r.toPush.map((e) => e.id), ['a']);
    expect(r.toWriteLocal, isEmpty);
  });

  test('규칙 2: 서버가 더 최신이면 로컬을 덮어쓴다', () {
    final r = planMerge(local: [local('a', t1, t1)], remote: [remote('a', t2)]);
    expect(r.toWriteLocal.map((e) => e.id), ['a']);
    expect(r.toPush, isEmpty);
  });

  test('규칙 2: 동률이면 서버가 이긴다', () {
    // 기기 시계 오차로 서로를 무한히 덮어쓰는 것을 막는다.
    final r = planMerge(local: [local('a', t1, t1)], remote: [remote('a', t1)]);
    expect(r.toWriteLocal.map((e) => e.id), ['a']);
    expect(r.toPush, isEmpty);
  });

  test('규칙 3: 서버에만 있으면 로컬에 넣는다', () {
    final r = planMerge(local: [], remote: [remote('b', t1)]);
    expect(r.toWriteLocal.map((e) => e.id), ['b']);
    expect(r.toDeleteLocal, isEmpty);
  });

  test('규칙 4: 올린 적 있는데 서버에 없으면 다른 기기에서 삭제된 것이다', () {
    final r = planMerge(local: [local('c', t1, t1)], remote: []);
    expect(r.toDeleteLocal, ['c']);
    expect(r.toPush, isEmpty);
  });

  test('네 규칙이 함께 동작한다', () {
    final r = planMerge(
      local: [local('unsynced', t2, null), local('older', t1, t1), local('gone', t1, t1)],
      remote: [remote('older', t2), remote('new', t1)],
    );
    expect(r.toPush.map((e) => e.id), ['unsynced']);
    expect(r.toWriteLocal.map((e) => e.id).toList()..sort(), ['new', 'older']);
    expect(r.toDeleteLocal, ['gone']);
  });

  test('양쪽이 비어 있으면 아무것도 하지 않는다', () {
    final r = planMerge(local: [], remote: []);
    expect(r.toWriteLocal, isEmpty);
    expect(r.toDeleteLocal, isEmpty);
    expect(r.toPush, isEmpty);
  });

  test('원격 레코드를 통째로 toWriteLocal에 싣는다', () {
    final doc = SyncRecord(
      id: 'a', updatedAt: t2, syncedAt: null,
      data: {'id': 'a', 'word': 'hi', 'review_count': 3},
    );
    final r = planMerge(local: [], remote: [doc]);
    expect(r.toWriteLocal.single.data, doc.data);
  });
}

File: test/merge.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { planMerge } from '../cloud/merge.js';

const T1 = '2026-08-01T00:00:00.000Z';
const T2 = '2026-08-02T00:00:00.000Z';

const local = (id, updated, synced) => ({ id, updated_at: updated, synced_at: synced });
const remote = (id, updated) => ({ id, updated_at: updated });

test('rule 1: an unsynced local item is pushed and never touched by the pull', () => {
  const r = planMerge([local('a', T1, null)], []);
  assert.deepEqual(r.toPush.map(i => i.id), ['a']);
  assert.deepEqual(r.toDeleteLocal, []);
  assert.deepEqual(r.toWriteLocal, []);
});

test('rule 1 holds even when the server has an older copy', () => {
  // The local edit has not been uploaded yet, so it must win regardless.
  const r = planMerge([local('a', T2, null)], [remote('a', T1)]);
  assert.deepEqual(r.toPush.map(i => i.id), ['a']);
  assert.deepEqual(r.toWriteLocal, []);
});

test('rule 1 takes precedence over rule 2: unsynced local older than remote must still push', () => {
  // Rule 1 says unsynced items must be pushed and never overwritten.
  // Rule 2 says newer timestamp wins. When they conflict, rule 1 wins.
  const r = planMerge([local('a', T1, null)], [remote('a', T2)]);
  assert.deepEqual(r.toPush.map(i => i.id), ['a']);
  assert.deepEqual(r.toWriteLocal, []);
  assert.deepEqual(r.toDeleteLocal, []);
});

test('rule 2: the newer updated_at wins — local newer means push', () => {
  const r = planMerge([local('a', T2, T1)], [remote('a', T1)]);
  assert.deepEqual(r.toPush.map(i => i.id), ['a']);
  assert.deepEqual(r.toWriteLocal, []);
});

test('rule 2: remote newer means overwrite local', () => {
  const r = planMerge([local('a', T1, T1)], [remote('a', T2)]);
  assert.deepEqual(r.toWriteLocal.map(i => i.id), ['a']);
  assert.deepEqual(r.toPush, []);
});

test('rule 2: a tie goes to the server', () => {
  // Clock skew between devices must not make them overwrite each other forever.
  const r = planMerge([local('a', T1, T1)], [remote('a', T1)]);
  assert.deepEqual(r.toWriteLocal.map(i => i.id), ['a']);
  assert.deepEqual(r.toPush, []);
});

test('rule 3: a remote-only item is inserted locally', () => {
  const r = planMerge([], [remote('b', T1)]);
  assert.deepEqual(r.toWriteLocal.map(i => i.id), ['b']);
  assert.deepEqual(r.toDeleteLocal, []);
});

test('rule 4: a synced local item missing from the server was deleted elsewhere', () => {
  const r = planMerge([local('c', T1, T1)], []);
  assert.deepEqual(r.toDeleteLocal, ['c']);
  assert.deepEqual(r.toPush, []);
});

test('handles all four rules together', () => {
  const r = planMerge(
    [local('unsynced', T2, null), local('older', T1, T1), local('gone', T1, T1)],
    [remote('older', T2), remote('new', T1)]
  );
  assert.deepEqual(r.toPush.map(i => i.id), ['unsynced']);
  assert.deepEqual(r.toWriteLocal.map(i => i.id).sort(), ['new', 'older']);
  assert.deepEqual(r.toDeleteLocal, ['gone']);
});

test('empty on both sides is a no-op', () => {
  assert.deepEqual(planMerge([], []), {
    toWriteLocal: [], toDeleteLocal: [], toPush: []
  });
});

test('carries the full remote record into toWriteLocal', () => {
  const doc = { id: 'a', updated_at: T2, word: 'hi', review_count: 3 };
  const r = planMerge([], [doc]);
  assert.deepEqual(r.toWriteLocal[0], doc);
});
